File: browserwncommon/src/main/java/org/sqlunet/browser/web/WebFragment.java
/*
 * Copyright (c) 2023. Bernard Bou
 */

package org.sqlunet.browser.web;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.Log;
import android.view.InflateException;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Toast;

import org.sqlunet.Pointer;
import org.sqlunet.bnc.sql.BncImplementation;
import org.sqlunet.browser.wn.DocumentTransformer;
import org.sqlunet.browser.wn.Settings;
import org.sqlunet.browser.wn.lib.BuildConfig;
import org.sqlunet.browser.wn.lib.R;
import org.sqlunet.dom.DomFactory;
import org.sqlunet.dom.DomTransformer;
import org.sqlunet.dom.DomValidator;
import org.sqlunet.provider.ProviderArgs;
import org.sqlunet.settings.LogUtils;
import org.sqlunet.settings.StorageSettings;
import org.sqlunet.sql.DataSource;
import org.sqlunet.sql.NodeFactory;
import org.sqlunet.wordnet.SensePointer;
import org.sqlunet.wordnet.SynsetPointer;
import org.sqlunet.wordnet.WordPointer;
import org.sqlunet.wordnet.sql.WordNetImplementation;
import org.w3c.dom.Document;

import java.net.URL;
import java.net.URLDecoder;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

/**
 * A fragment representing a SqlUNet web view.
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
public class WebFragment extends Fragment
{
	static private final String TAG = "WebF";

	static private final String SQLUNET_NS = "http://org.sqlunet";

	class WebDocumentStringLoader implements DocumentStringLoader
	{
		@NonNull
		final Context context;
		final private int type;
		final private String data;
		final private Parcelable pointer;
		final private Character pos;
		final private int sources;
		final private boolean xml;

		WebDocumentStringLoader(@NonNull final Context context, final int type, final String data, final Parcelable pointer, final Character pos, final int sources, final boolean xml)
		{
			super();
			this.context = context;
			this.data = data;
			this.type = type;
			this.pointer = pointer;
			this.pos = pos;
			this.sources = sources;
			this.xml = xml;
		}

		@Nullable
		@SuppressWarnings({"boxing"})
		@Override
		public String getDoc()
		{
			try (DataSource dataSource = new DataSource(StorageSettings.getDatabasePath(this.context)))
			{
				// data source
				final SQLiteDatabase db = dataSource.getConnection();
				WordNetImplementation.init(db);

				// dom documents
				Document wnDomDoc = null;
				Document bncDomDoc = null;

				// selector mode
				final boolean isSelector = this.data != null;

				// make documents
				if (isSelector)
				{
					// this is a selector query
					if (Settings.Source.WORDNET.test(this.sources))
					{
						wnDomDoc = new WordNetImplementation().querySelectorDoc(db, this.data);
					}
				}
				else
				{
					// this is a detail query
					switch (this.type)
					{
						case ProviderArgs.ARG_QUERYTYPE_ALL:
							if (pointer != null)
							{
								final SensePointer sense2Pointer = (SensePointer) this.pointer;
								final long wordId = sense2Pointer.getWordId();
								final Long synsetId = sense2Pointer.getSynsetId();
								if (Settings.Source.WORDNET.test(this.sources))
								{
									wnDomDoc = new WordNetImplementation().queryDoc(db, wordId, synsetId, true, false);
								}
								if (Settings.Source.BNC.test(this.sources))
								{
									bncDomDoc = new BncImplementation().queryDoc(db, wordId, pos);
								}
							}
							break;

						case ProviderArgs.ARG_QUERYTYPE_WORD:
							@SuppressWarnings("TypeMayBeWeakened") final WordPointer wordPointer = (WordPointer) this.pointer;
							Log.d(TAG, "word=" + wordPointer);
							if (wordPointer != null && Settings.Source.WORDNET.test(this.sources))
							{
								wnDomDoc = new WordNetImplementation().queryWordDoc(db, wordPointer.getWordId());
							}
							break;

						case ProviderArgs.ARG_QUERYTYPE_SYNSET:
							@SuppressWarnings("TypeMayBeWeakened") final SynsetPointer synsetPointer = (SynsetPointer) this.pointer;
							Log.d(TAG, "synset=" + synsetPointer);
							if (synsetPointer != null && Settings.Source.WORDNET.test(this.sources))
							{
								wnDomDoc = new WordNetImplementation().querySynsetDoc(db, synsetPointer.getSynsetId());
							}
							break;
					}
				}

				// stringify
				return docsToString(data, xml, isSelector, wnDomDoc, bncDomDoc);
			}
			catch (@NonNull final Exception e)
			{
				Log.e(TAG, "getDoc", e);
			}
			return null;
		}
	}

	/**
	 * HTML stuff
	 */
	static private final String BODY1 = "<HTML><HEAD>";
	static private final String BODY2 = "</HEAD><BODY>";
	static private final String BODY3 = "</BODY></HTML>";
	static private final String TOP = "<DIV class='titlesection'><IMG class='titleimg' src='images/logo.png'/></DIV>";
	static private final String STYLESHEET1 = "<LINK rel='stylesheet' type='text/css' href='";
	static private final String STYLESHEET2 = "' />";
	static private final String SCRIPT1 = "<SCRIPT type='text/javascript' src='";
	static private final String SCRIPT2 = "'></SCRIPT>";
	static private final String LIST1 = "<UL style='display: block;'>";
	static private final String LIST2 = "</UL>";
	static private final String ITEM1 = "<LI class='treeitem treepanel'>";
	static private final String ITEM2 = "</LI>";

	/**
	 * WebView
	 */
	private WebView webview;

	/**
	 * View model
	 */
	private WebModel model;

	/**
	 * Mandatory empty constructor for the fragment manager to instantiate the fragment (e.g. upon screen orientation changes).
	 */
	public WebFragment()
	{
	}

	@Override
	public View onCreateView(@NonNull final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState)
	{
		try
		{
			// view
			final View view = inflater.inflate(R.layout.fragment_web, container, false);

			// webview
			this.webview = view.findViewById(R.id.webView);

			return view;
		}
		catch (InflateException e)
		{
			Toast.makeText(requireContext(), "No WebView support", Toast.LENGTH_LONG).show();
			return null;
		}
	}

	@Override
	public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState)
	{
		super.onViewCreated(view, savedInstanceState);

		// models
		makeModels();

		// load data
		load();
	}

	/**
	 * Make view models
	 */
	private void makeModels()
	{
		final boolean xml = Settings.getXmlPref(requireContext());
		this.model = new ViewModelProvider(this).get("wn:web(doc)", WebModel.class);
		this.model.getData().observe(getViewLifecycleOwner(), doc -> {
			Log.d(TAG, "onLoadFinished");
			final String mimeType = xml ? "text/xml" : "text/html";
			final String baseUrl = "file:///android_asset/";
			WebFragment.this.webview.loadDataWithBaseURL(baseUrl, doc, mimeType, "utf-8", null);
			//WebFragment.this.webview.loadUrl("_about:blank");
		});
	}

	/**
	 * Load web view with data
	 */
	@SuppressWarnings("UnusedReturnValue")
	@SuppressLint("SetJavaScriptEnabled")
	private void load()
	{
		// settings
		final WebSettings webSettings = this.webview.getSettings();

		// enable javascript
		webSettings.setJavaScriptEnabled(true);

		// zoom support
		webSettings.setSupportZoom(true);
		webSettings.setBuiltInZoomControls(true);

		// client
		final WebViewClient webClient = new WebViewClient()
		{
			@Override
			public boolean shouldOverrideUrlLoading(@NonNull final WebView view, @NonNull final String urlString)
			{
				final Uri uri = Uri.parse(urlString);
				return handleUrl(uri);
			}

			@TargetApi(Build.VERSION_CODES.N)
			@Override
			public boolean shouldOverrideUrlLoading(@NonNull final WebView view, @NonNull final WebResourceRequest request)
			{
				final Uri uri = request.getUrl();
				return handleUrl(uri);
			}

			private boolean handleUrl(@NonNull final Uri uri)
			{
				Log.d(TAG, "Uri " + uri);
				try
				{
					final String query = URLDecoder.decode(uri.getQuery(), "UTF-8");
					final String[] target = query.split("=");
					final String name = target[0];
					final String value = target[1];
					Log.d(TAG, "QUERY " + query + " name=" + name + " value=" + value);
					final Intent targetIntent = new Intent(requireContext(), WebActivity.class);
					if ("word".equals(name)) //
					{
						targetIntent.putExtra(ProviderArgs.ARG_QUERYSTRING, value);
					}
					else
					{
						final long id = Long.parseLong(value);

						// warn with id
						final Activity activity = getActivity();
						if (activity != null && !isDetached() && !activity.isFinishing() && !activity.isDestroyed())
						{
							activity.runOnUiThread(() -> Toast.makeText(activity, "id=" + id, Toast.LENGTH_SHORT).show());
						}

						// prepare data
						int type;
						Pointer pointer;

						switch (name)
						{
							case "wordid":
								type = ProviderArgs.ARG_QUERYTYPE_WORD;
								pointer = new WordPointer(id);
								break;
							case "synsetid":
								type = ProviderArgs.ARG_QUERYTYPE_SYNSET;
								pointer = new SynsetPointer(id);
								break;
							default:
								Log.e(TAG, "Ill-formed Uri: " + uri);
								return false;
						}

						final int recurse = Settings.getRecursePref(requireContext());
						targetIntent.putExtra(ProviderArgs.ARG_QUERYTYPE, type);
						targetIntent.putExtra(ProviderArgs.ARG_QUERYPOINTER, pointer);
						targetIntent.putExtra(ProviderArgs.ARG_QUERYRECURSE, recurse);
					}
					targetIntent.setAction(ProviderArgs.ACTION_QUERY);
					startActivity(targetIntent);
					return true;
				}
				catch (@NonNull final Exception e)
				{
					Log.e(TAG, "Error while loading Uri: " + uri, e);
				}
				return false;
			}
		};
		this.webview.setWebViewClient(webClient);

		// settings sources
		Context context = requireContext();

		int mask = 0;
		if (Settings.getWordNetPref(context))
		{
			mask = Settings.Source.WORDNET.set(mask);
		}
		if (Settings.getBncPref(context))
		{
			mask |= Settings.Source.BNC.set(mask);
		}
		final int sources = mask;

		// settings output
		final boolean xml = Settings.getXmlPref(context);

		// unmarshal arguments
		Bundle args = getArguments();
		assert args != null;

		// type
		final int type = args.getInt(ProviderArgs.ARG_QUERYTYPE, ProviderArgs.ARG_QUERYTYPE_ALL);

		// pointer
		final Parcelable pointer = android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.TIRAMISU ? args.getParcelable(ProviderArgs.ARG_QUERYPOINTER, Parcelable.class) : args.getParcelable(ProviderArgs.ARG_QUERYPOINTER);
		Log.d(TAG, "query=" + pointer);

		// hint
		final String posString = args.getString(ProviderArgs.ARG_HINTPOS);
		final Character pos = posString == null ? null : posString.charAt(0);

		// text
		final String data = args.getString(ProviderArgs.ARG_QUERYSTRING);
		Log.d(TAG, "data=" + data);

		// load the contents
		this.model.loadData(new WebDocumentStringLoader(requireContext(), type, data, pointer, pos, sources, xml));
	}

	/**
	 * Assemble result
	 *
	 * @param word       query word
	 * @param xml        assemble as xml (or document to be xslt-transformed if false)
	 * @param isSelector is selector source
	 * @param wnDomDoc   wordnet document
	 * @param bncDomDoc  bnc document
	 * @return string
	 */
	private String docsToString(@SuppressWarnings("UnusedParameters") final String word,  //
			final boolean xml,  //
			final boolean isSelector,  //
			@Nullable final Document wnDomDoc,  //
			@Nullable final Document bncDomDoc)
	{
		// LogUtils.writeLog(DomTransformer.docToXml(wnDomDoc), false, "wn_sqlunet.log");
		// LogUtils.writeLog(DomTransformer.docToXml(bncDomDoc), false, "bnc_sqlunet.log");

		String data;
		if (xml)
		{
			// merge all into one
			final Document rootDomDoc = DomFactory.makeDocument();
			NodeFactory.makeRootNode(rootDomDoc, rootDomDoc, "sqlunet", null, WebFragment.SQLUNET_NS);
			if (wnDomDoc != null)
			{
				rootDomDoc.getDocumentElement().appendChild(rootDomDoc.importNode(wnDomDoc.getFirstChild(), true));
			}
			if (bncDomDoc != null)
			{
				rootDomDoc.getDocumentElement().appendChild(rootDomDoc.importNode(bncDomDoc.getFirstChild(), true));
			}

			data = DomTransformer.docToXml(rootDomDoc);
			if (BuildConfig.DEBUG)
			{
				LogUtils.writeLog(data, false, requireContext(),null);
				final URL xsd = DocumentTransformer.class.getResource("/org/sqlunet/SqlUNet.xsd");
				assert xsd != null;
				DomValidator.validateStrings(xsd, data);
				Log.d(TAG, "output=\n" + data);
			}
		}
		else
		{
			final StringBuilder sb = new StringBuilder();

			// header
			sb.append(BODY1);

			// css style sheet
			sb.append(STYLESHEET1).append("css/style.css").append(STYLESHEET2);
			sb.append(STYLESHEET1).append("css/tree.css").append(STYLESHEET2);
			sb.append(STYLESHEET1).append("css/wordnet.css").append(STYLESHEET2);
			if (bncDomDoc != null)
			{
				sb.append(STYLESHEET1).append("css/bnc.css").append(STYLESHEET2);
			}

			// javascripts
			sb.append(SCRIPT1).append("js/tree.js").append(SCRIPT2);
			sb.append(SCRIPT1).append("js/sarissa.js").append(SCRIPT2);
			sb.append(SCRIPT1).append("js/ajax.js").append(SCRIPT2);
			sb.append(SCRIPT1).append("js/wordnet.js").append(SCRIPT2);
			if (bncDomDoc != null)
			{
				sb.append(SCRIPT1).append("js/bnc.js").append(SCRIPT2);
			}

			// body
			sb.append(BODY2);

			// top
			sb.append(TOP);

			// xslt-transformed data
			sb.append(LIST1);
			if (wnDomDoc != null)
			{
				sb.append(ITEM1);
				sb.append(new DocumentTransformer().docToHtml(wnDomDoc, Settings.Source.WORDNET.toString(), isSelector));
				sb.append(ITEM2);
			}
			if (bncDomDoc != null)
			{
				sb.append(ITEM1);
				sb.append(new DocumentTransformer().docToHtml(bncDomDoc, Settings.Source.BNC.toString(), isSelector));
				sb.append(ITEM2);
			}
			sb.append(LIST2);

			// tail
			sb.append(BODY3);

			data = sb.toString();

			if (BuildConfig.DEBUG)
			{
				final URL xsd = DocumentTransformer.class.getResource("/org/sqlunet/SqlUNet.xsd");
				assert xsd != null;
				DomValidator.validateDocs(xsd, wnDomDoc, bncDomDoc);
				LogUtils.writeLog(false, requireContext(),null, wnDomDoc, bncDomDoc);
				LogUtils.writeLog(data, false, requireContext(),null);
				Log.d(TAG, "output=\n" + data);
			}
		}
		return data;
	}
}

File: browservn/src/main/java/org/sqlunet/browser/vn/web/WebFragment.kt
/*
 * Copyright (c) 2023. Bernard Bou <1313ou@gmail.com>
 */
package org.sqlunet.browser.vn.web

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import android.view.InflateException
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import org.sqlunet.Pointer
import org.sqlunet.browser.getParcelable
import org.sqlunet.browser.vn.BuildConfig
import org.sqlunet.browser.vn.DocumentTransformer
import org.sqlunet.browser.vn.R
import org.sqlunet.browser.web.DocumentStringLoader
import org.sqlunet.browser.web.WebModel
import org.sqlunet.dom.DomFactory.makeDocument
import org.sqlunet.dom.DomTransformer.docToXml
import org.sqlunet.dom.DomValidator.validateDocs
import org.sqlunet.dom.DomValidator.validateStrings
import org.sqlunet.propbank.PbRoleSetPointer
import org.sqlunet.propbank.sql.PropBankImplementation
import org.sqlunet.provider.ProviderArgs
import org.sqlunet.settings.LogUtils.writeLog
import org.sqlunet.settings.Settings
import org.sqlunet.settings.StorageSettings
import org.sqlunet.sql.DataSource
import org.sqlunet.sql.NodeFactory.makeRootNode
import org.sqlunet.verbnet.VnClassPointer
import org.sqlunet.verbnet.sql.VerbNetImplementation
import org.sqlunet.wordnet.SensePointer
import org.sqlunet.wordnet.SynsetPointer
import org.sqlunet.wordnet.WordPointer
import org.sqlunet.wordnet.sql.WordNetImplementation
import org.sqlunet.wordnet.sql.WordNetImplementation.Companion.init
import org.w3c.dom.Document
import java.net.URLDecoder

/**
 * A fragment representing a SqlUNet web view.
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
class WebFragment : Fragment() {

    private inner class WebDocumentStringLoader(val context: Context, val pointer: Parcelable?, val pos: Char, val type: Int, val data: String?, val sources: Int, val xml: Boolean) : DocumentStringLoader {

        override fun getDoc(): String? {
            try {
                DataSource(StorageSettings.getDatabasePath(context)).use {

                    // data source
                    val db = it.connection
                    init(db)

                    // dom documents
                    var wnDomDoc: Document? = null
                    var vnDomDoc: Document? = null
                    var pbDomDoc: Document? = null

                    // selector mode
                    val isSelector = data != null

                    // make documents
                    if (isSelector) {
                        // this is a selector query
                        if (org.sqlunet.browser.vn.VnSettings.Source.WORDNET.test(sources)) {
                            wnDomDoc = WordNetImplementation().querySelectorDoc(db, data!!)
                        }
                        if (org.sqlunet.browser.vn.VnSettings.Source.VERBNET.test(sources)) {
                            vnDomDoc = VerbNetImplementation().querySelectorDoc(db, data!!)
                        }
                        if (org.sqlunet.browser.vn.VnSettings.Source.PROPBANK.test(sources)) {
                            pbDomDoc = PropBankImplementation().querySelectorDoc(db, data!!)
                        }
                    } else {
                        // this is a detail query
                        when (type) {
                            ProviderArgs.ARG_QUERYTYPE_ALL -> if (pointer != null) {
                                if (pointer is org.sqlunet.browser.vn.xselector.XSelectorPointer) {
                                    val xPointer = pointer
                                    val xSources = xPointer.getXSources()
                                    val xClassId = xPointer.getXClassId()
                                    // var xMemberId = xpointer.getXMemberId()
                                    val wordId = xPointer.getWordId()
                                    val synsetId = xPointer.getSynsetId()
                                    if (xSources == null || xSources.contains("wn")) {
                                        wnDomDoc = WordNetImplementation().querySenseDoc(db, wordId, synsetId)
                                    }
                                    if ((xSources == null || xSources.contains("vn")) && xClassId != null) {
                                        vnDomDoc = VerbNetImplementation().queryClassDoc(db, xClassId, pos)
                                    }
                                    if ((xSources == null || xSources.contains("pb")) && xClassId != null) {
                                        pbDomDoc = PropBankImplementation().queryRoleSetDoc(db, xClassId, pos)
                                    }
                                } else {
                                    val sense2Pointer = pointer as SensePointer
                                    val wordId = sense2Pointer.getWordId()
                                    val synsetId = sense2Pointer.getSynsetId()
                                    if (org.sqlunet.browser.vn.VnSettings.Source.WORDNET.test(sources)) {
                                        wnDomDoc = WordNetImplementation().queryDoc(db, wordId, synsetId, withRelations = true, recurse = false)
                                    }
                                    if (org.sqlunet.browser.vn.VnSettings.Source.VERBNET.test(sources)) {
                                        vnDomDoc = VerbNetImplementation().queryDoc(db, wordId, synsetId, pos)
                                    }
                                    if (org.sqlunet.browser.vn.VnSettings.Source.PROPBANK.test(sources)) {
                                        pbDomDoc = PropBankImplementation().queryDoc(db, wordId, pos)
                                    }
                                }
                            }

                            ProviderArgs.ARG_QUERYTYPE_WORD -> {
                                val wordPointer = pointer as WordPointer?
                                Log.d(TAG, "ArgPosition: word=$wordPointer")
                                if (wordPointer != null && org.sqlunet.browser.vn.VnSettings.Source.WORDNET.test(sources)) {
                                    wnDomDoc = WordNetImplementation().queryWordDoc(db, wordPointer.getWordId())
                                }
                            }

                            ProviderArgs.ARG_QUERYTYPE_SYNSET -> {
                                val synsetPointer = pointer as SynsetPointer?
                                Log.d(TAG, "ArgPosition: synset=$synsetPointer")
                                if (synsetPointer != null && org.sqlunet.browser.vn.VnSettings.Source.WORDNET.test(sources)) {
                                    wnDomDoc = WordNetImplementation().querySynsetDoc(db, synsetPointer.getSynsetId())
                                }
                            }

                            ProviderArgs.ARG_QUERYTYPE_VNCLASS -> {
                                val vnclassPointer = pointer as VnClassPointer?
                                Log.d(TAG, "ArgPosition: vnclass=$vnclassPointer")
                                if (vnclassPointer != null && org.sqlunet.browser.vn.VnSettings.Source.VERBNET.test(sources)) {
                                    vnDomDoc = VerbNetImplementation().queryClassDoc(db, vnclassPointer.id, null)
                                }
                            }

                            ProviderArgs.ARG_QUERYTYPE_PBROLESET -> {
                                val pbroleSetPointer = pointer as PbRoleSetPointer?
                                Log.d(TAG, "ArgPosition: pbroleset=$pbroleSetPointer")
                                if (pbroleSetPointer != null && org.sqlunet.browser.vn.VnSettings.Source.PROPBANK.test(sources)) {
                                    pbDomDoc = PropBankImplementation().queryRoleSetDoc(db, pbroleSetPointer.id, null)
                                }
                            }
                        }
                    }

                    // stringify
                    return docsToString(data, xml, isSelector, wnDomDoc, vnDomDoc, pbDomDoc)
                }
            } catch (e: Exception) {
                Log.e(TAG, "getDoc", e)
            }
            return null
        }
    }

    /**
     * WebView
     */
    private var webview: WebView? = null

    // View model
    private var model: WebModel? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return try {
            inflater.inflate(R.layout.fragment_web, container, false)
        } catch (e: InflateException) {
            Toast.makeText(requireContext(), "No WebView support", Toast.LENGTH_LONG).show()
            null
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // webview
        webview = view.findViewById(R.id.webView)

        // models
        makeModels()

        // load data
        load()
    }

    /**
     * Make view models
     */
    private fun makeModels() {
        val xml: Boolean = Settings.getXmlPref(requireContext())
        model = ViewModelProvider(this)["vn:web(doc)", WebModel::class.java]
        model!!.getData().observe(getViewLifecycleOwner()) { doc: String? ->
            Log.d(TAG, "onLoadFinished")
            val mimeType = if (xml) "text/xml" else "text/html"
            val baseUrl = "file:///android_asset/"
            webview!!.loadDataWithBaseURL(baseUrl, doc!!, mimeType, "utf-8", null)
        }
    }

    /**
     * Load web view with data
     */
    @SuppressLint("SetJavaScriptEnabled")
    private fun load() {
        // settings
        val webSettings = webview!!.getSettings()

        // enable javascript
        webSettings.javaScriptEnabled = true

        // zoom support
        webSettings.setSupportZoom(true)
        webSettings.builtInZoomControls = true

        // client
        val webClient: WebViewClient = object : WebViewClient() {

            @Deprecated("Deprecated in Java")
            override fun shouldOverrideUrlLoading(view: WebView, urlString: String): Boolean {
                val uri = Uri.parse(urlString)
                return handleUrl(uri)
            }

            @TargetApi(Build.VERSION_CODES.N)
            override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
                val uri = request.url
                return handleUrl(uri)
            }

            private fun handleUrl(uri: Uri): Boolean {
                Log.d(TAG, "Uri $uri")
                try {
                    val query = URLDecoder.decode(uri.query, "UTF-8")
                    val target = query.split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                    val name = target[0]
                    val value = target[1]
                    Log.d(TAG, "Query: $query name=$name value=$value")
                    val targetIntent = Intent(requireContext(), WebActivity::class.java)
                    if ("word" == name) {
                        targetIntent.putExtra(ProviderArgs.ARG_QUERYSTRING, value)
                    } else {
                        val id = value.toLong()

                        // warn with id
                        val activity: Activity? = activity
                        if (activity != null && !isDetached && !activity.isFinishing && !activity.isDestroyed) {
                            activity.runOnUiThread(Runnable { Toast.makeText(activity, "id=$id", Toast.LENGTH_SHORT).show() })
                        }

                        // prepare data
                        val type: Int
                        val pointer: Pointer
                        when (name) {
                            "vnclassid" -> {
                                type = ProviderArgs.ARG_QUERYTYPE_VNCLASS
                                pointer = VnClassPointer(id)
                            }

                            "pbrolesetid" -> {
                                type = ProviderArgs.ARG_QUERYTYPE_PBROLESET
                                pointer = PbRoleSetPointer(id)
                            }

                            "wordid" -> {
                                type = ProviderArgs.ARG_QUERYTYPE_WORD
                                pointer = WordPointer(id)
                            }

                            "synsetid" -> {
                                type = ProviderArgs.ARG_QUERYTYPE_SYNSET
                                pointer = SynsetPointer(id)
                            }

                            else -> {
                                Log.e(TAG, "Ill-formed Uri: $uri")
                                return false
                            }
                        }
                        targetIntent.putExtra(ProviderArgs.ARG_QUERYTYPE, type)
                        targetIntent.putExtra(ProviderArgs.ARG_QUERYPOINTER, pointer)
                    }
                    targetIntent.setAction(ProviderArgs.ACTION_QUERY)
                    startActivity(targetIntent)
                    return true
                } catch (e: Exception) {
                    Log.e(TAG, "Error while loading Uri: $uri", e)
                }
                return false
            }
        }
        webview!!.setWebViewClient(webClient)

        // context
        val context = requireContext()

        // settings sources
        val enable = org.sqlunet.browser.vn.VnSettings.getAllPref(context)

        // settings output
        val xml: Boolean = Settings.getXmlPref(context)

        // unmarshal arguments
        val args = requireArguments()

        // type
        val type = args.getInt(ProviderArgs.ARG_QUERYTYPE, ProviderArgs.ARG_QUERYTYPE_ALL)

        // pointer
        val pointer = getParcelable(args, ProviderArgs.ARG_QUERYPOINTER)
        Log.d(TAG, "ArgPosition: query=$pointer")

        // hint
        val posString = args.getString(ProviderArgs.ARG_HINTPOS)
        val pos = posString?.get(0)!!

        // text
        val data = args.getString(ProviderArgs.ARG_QUERYSTRING)
        Log.d(TAG, "ArgPosition: data=$data")

        // load the contents
        model!!.loadData(WebDocumentStringLoader(context, pointer, pos, type, data, enable, xml))
    }

    /**
     * Assemble result
     *
     * @param word       query word
     * @param xml        assemble as xml (or document to be xslt-transformed if false)
     * @param isSelector is selector source
     * @param wnDomDoc   wordnet document
     * @param vnDomDoc   verbnet document
     * @param pbDomDoc   propbank document
     * @return string
     */
    private fun docsToString(
        word: String?,
        xml: Boolean,
        isSelector: Boolean,
        wnDomDoc: Document?,
        vnDomDoc: Document?,
        pbDomDoc: Document?
    ): String {
        // LogUtils.writeLog(DomTransformer.docToXml(wnDomDoc), false, "wn_sqlunet.log")
        // LogUtils.writeLog(DomTransformer.docToXml(vnDomDoc), false, "vn_sqlunet.log")
        // LogUtils.writeLog(DomTransformer.docToXml(pbDomDoc), false, "pb_sqlunet.log")
        // LogUtils.writeLog(DomTransformer.docToXml(bncDomDoc), false, "bnc_sqlunet.log")
        val data: String
        if (xml) {
            // merge all into one
            val rootDomDoc = makeDocument()
            makeRootNode(rootDomDoc, rootDomDoc, "sqlunet", null, SQLUNET_NS)
            if (wnDomDoc != null) {
                rootDomDoc.documentElement.appendChild(rootDomDoc.importNode(wnDomDoc.firstChild, true))
            }
            if (vnDomDoc != null) {
                rootDomDoc.documentElement.appendChild(rootDomDoc.importNode(vnDomDoc.firstChild, true))
            }
            if (pbDomDoc != null) {
                rootDomDoc.documentElement.appendChild(rootDomDoc.importNode(pbDomDoc.firstChild, true))
            }
            data = docToXml(rootDomDoc)
            if (BuildConfig.DEBUG) {
                writeLog(data, false, requireContext(), null)
                val xsd = DocumentTransformer::class.java.getResource("/org/sqlunet/SqlUNet.xsd")!!
                validateStrings(xsd, data)
                Log.d(TAG, "output=\n$data")
            }
        } else {
            val sb = StringBuilder()

            // header
            sb.append(BODY1)

            // css style sheet
            sb.append(STYLESHEET1).append("css/style.css").append(STYLESHEET2)
            sb.append(STYLESHEET1).append("css/tree.css").append(STYLESHEET2)
            sb.append(STYLESHEET1).append("css/wordnet.css").append(STYLESHEET2)
            if (vnDomDoc != null) {
                sb.append(STYLESHEET1).append("css/verbnet.css").append(STYLESHEET2)
            }
            if (pbDomDoc != null) {
                sb.append(STYLESHEET1).append("css/propbank.css").append(STYLESHEET2)
            }

            // javascripts
            sb.append(SCRIPT1).append("js/tree.js").append(SCRIPT2)
            sb.append(SCRIPT1).append("js/sarissa.js").append(SCRIPT2)
            sb.append(SCRIPT1).append("js/ajax.js").append(SCRIPT2)
            sb.append(SCRIPT1).append("js/wordnet.js").append(SCRIPT2)
            if (vnDomDoc != null) {
                sb.append(SCRIPT1).append("js/verbnet.js'></script>")
            }
            if (pbDomDoc != null) {
                sb.append(SCRIPT1).append("js/propbank.js").append(SCRIPT2)
            }

            // body
            sb.append(BODY2)

            // top
            sb.append(TOP)

            // xslt-transformed data
            sb.append(LIST1)
            if (vnDomDoc != null) {
                sb.append(ITEM1)
                sb.append(DocumentTransformer().docToHtml(vnDomDoc, org.sqlunet.browser.vn.VnSettings.Source.VERBNET.toString(), isSelector))
                sb.append(ITEM2)
            }
            if (pbDomDoc != null) {
                sb.append(ITEM1)
                sb.append(DocumentTransformer().docToHtml(pbDomDoc, org.sqlunet.browser.vn.VnSettings.Source.PROPBANK.toString(), isSelector))
                sb.append(ITEM2)
            }
            if (wnDomDoc != null) {
                sb.append(ITEM1)
                sb.append(DocumentTransformer().docToHtml(wnDomDoc, org.sqlunet.browser.vn.VnSettings.Source.WORDNET.toString(), isSelector))
                sb.append(ITEM2)
            }
            sb.append(LIST2)

            // tail
            sb.append(BODY3)
            data = sb.toString()
            if (BuildConfig.DEBUG) {
                val xsd = DocumentTransformer::class.java.getResource("/org/sqlunet/SqlUNet.xsd")!!
                validateDocs(xsd, wnDomDoc!!, vnDomDoc!!, pbDomDoc!!)
                writeLog(false, requireContext(), null, wnDomDoc, vnDomDoc, pbDomDoc)
                writeLog(data, false, requireContext(), null)
                Log.d(TAG, "output=\n$data")
            }
        }
        return data
    }

    companion object {

        private const val TAG = "WebF"
        const val FRAGMENT_TAG = "web"
        private const val SQLUNET_NS = "http://org.sqlunet"

        /**
         * HTML stuff
         */
        private const val BODY1 = "<HTML><HEAD>"
        private const val BODY2 = "</HEAD><BODY>"
        private const val BODY3 = "</BODY></HTML>"
        private const val TOP = "<DIV class='titlesection'><IMG class='titleimg' src='images/logo.png'/></DIV>"
        private const val STYLESHEET1 = "<LINK rel='stylesheet' type='text/css' href='"
        private const val STYLESHEET2 = "' />"
        private const val SCRIPT1 = "<SCRIPT type='text/javascript' src='"
        private const val SCRIPT2 = "'></SCRIPT>"
        private const val LIST1 = "<UL style='display: block;'>"
        private const val LIST2 = "</UL>"
        private const val ITEM1 = "<LI class='treeitem treepanel'>"
        private const val ITEM2 = "</LI>"
    }
}
